### PR TITLE
Mobile: Fixes #9204: Handle callback url triggered app launch

### LINKED ISSUE #9204: Mobile (iOS/Android) implement external-links/x-callback-url
Making an issue for the mobile part of #5168. To my knowledge this has not yet been implemented on any mobile platform. It seems there are a number of other bug reports that mention this, without a successful fix. https://github.com/laurent22/joplin/issues?q=is%3Aissue+x-callback & https://discourse.joplinapp.org/search?q=x-callback

Implementing this feature allows for NFC tags to be used on mobile devices with Joplin.

## Environment

Joplin version: Latest (iOS 12.12.3)
Platforms:  iOS 17.0.3 and iOS 15.7.1 (and Android from what I've read, but I haven't tested).

## Steps to reproduce (iOS)
1. In app, copy the markdown link for the note. eg: `[My Test Note](:/1586e8e9da1e495eb08c7b4cdf1a6f50)`
2. Modify it to a proper external link and copy. eg: `joplin://x-callback-url/openNote?id=1586e8e9da1e495eb08c7b4cdf1a6f50`
3. Switch to a different note
4. [Optional] Close the app
5. Paste the external link into safari or other browser
6. Joplin launches, but not to the correct note

## Describe what you expected to happen
Step 6 above should launch to the correct note.
Extra: UI entry to copy the external link, similar to the markdown link in the app. Currently desktop app has this

## Logfile
No relevant entries in the logfile

#### Relevant
https://reactnative.dev/docs/linking?syntax=ios

### PR DESCRIPTION
**Pull request by tiberiusteng**
Existing `Linking.addEventListener` will only receive URLs triggered after app launch, adding `Linking.getInitialURL` during app start to get the URL triggered app launch (if there's one).

Compiled and tested on Fairphone 4 with both biometrics disabled and enabled:

- Force stop app
- Scan QR code from the system-provided tile in notification area, touch 'Open with Joplin'
- Do biometrics authorization
- Confirm opened linked note

Latest commit f4ee4a178e692dfc7fbe60afed6f7993652542da can't build due to `shim.attachFileToNote` signature change, so I tested with ff09937ca299541b37303afc138cd1af6b6e2b03.

### CODE REVIEW

**Comment by tiberiusteng** [comment]
Fixes #9204

File: packages/app-mobile/root.tsx [hunk 1]
```diff
@@ -948,6 +948,12 @@ class AppComponent extends React.Component {
 				throw error;
 			}
 
+			const url = await Linking.getInitialURL();
+			if (url !== null && isCallbackUrl(url)) {
```
Comment (laurent22) [inline packages/app-mobile/root.tsx]:
```suggestion
			if (url && isCallbackUrl(url)) {
```

File: packages/app-mobile/root.tsx [hunk 2]
```diff
@@ -948,6 +948,12 @@ class AppComponent extends React.Component {
 				throw error;
 			}
 
+			const url = await Linking.getInitialURL();
```
Comment (laurent22) [inline packages/app-mobile/root.tsx:960]: Could you add a comment explaining why we need that here please?

**Comment by laurent22** [comment]
Thanks for fixing this @tiberiusteng!

**Comment by tiberiusteng** [comment]
Hi @laurent22, I've tried with latest Android versions (3.1.7) and looks like this patch is not included in current android release, is this intended? Am I correct to assume this will be included in 3.2 release? Thanks!

**Pull request merged**